index-scheduler/src/batch.rs

```diff
@@ -816,49 +820,56 @@ impl IndexScheduler {
     /// Delete each given task from all the databases (if it is deleteable).
     ///
     /// Return the number of tasks that were actually deleted
-    fn delete_matched_tasks(&self, wtxn: &mut RwTxn, matched_tasks: &[u32]) -> Result<usize> {
+    fn delete_matched_tasks(
+        &self,
+        wtxn: &mut RwTxn,
+        matched_tasks: &RoaringBitmap,
+    ) -> Result<usize> {
         // 1. Remove from this list the tasks that we are not allowed to delete
         let enqueued_tasks = self.get_status(wtxn, Status::Enqueued)?;
 
         let processing_tasks = &self.processing_tasks.read().unwrap().1;
-        let to_delete_tasks = matched_tasks
-            .iter()
-            .filter(|&&task_id| {
-                !processing_tasks.contains(task_id) && !enqueued_tasks.contains(task_id)
-            })
-            .copied();
-        let to_delete_tasks = RoaringBitmap::from_iter(to_delete_tasks);
-        // 2. We now have a list of tasks to delete, delete them
 
-        for task_id in to_delete_tasks.iter() {
-            let task = self.all_tasks.get(wtxn, &BEU32::new(task_id))?.unwrap();
-            self.delete_task(wtxn, &task)?;
-        }
+        // TODO: Lo: Take the intersection of `matched_tasks` and `all_tasks` first,
+        // so that we end up with the correct count for `deleted_tasks` (the value returned
+        // by this function). Related snapshot test:
+        // `task_deletion_delete_same_task_twice/task_deletion_processed.snap`
+        let mut to_delete_tasks = matched_tasks - processing_tasks;
+        to_delete_tasks -= enqueued_tasks;
 
-        Ok(to_delete_tasks.len() as usize)
-    }
+        // 2. We now have a list of tasks to delete, delete them
 
-    /// Delete the given task from all the databases
-    fn delete_task(&self, wtxn: &mut RwTxn, task: &Task) -> Result<()> {
-        let task_id = BEU32::new(task.uid);
+        let mut affected_indexes = HashSet::new();
+        let mut affected_statuses = HashSet::new();
+        let mut affected_kinds = HashSet::new();
 
-        if let Some(indexes) = task.indexes() {
-            for index in indexes {
-                self.update_index(wtxn, index, |bitmap| {
-                    bitmap.remove(task.uid);
-                })?;
+        for task_id in to_delete_tasks.iter() {
+            if let Some(task) = self.get_task(wtxn, task_id)? {
+                if let Some(task_indexes) = task.indexes() {
+                    affected_indexes.extend(task_indexes.into_iter().map(|x| x.to_owned()));
+                }
+                affected_statuses.insert(task.status);
+                affected_kinds.insert(task.kind.as_kind());
```

Comment:
You should write an else stmt where you effectively count how may tasks has been deleted

Reply:
ouch, yep, I meant to do that but completely forgot :/